cards4-template.js

```diff
@@ -21,13 +50,28 @@ var makeCard = // receive factory with external name `makeCard`
 //-----------------------
 // Instance Methods:
 //-----------------------
-
-    function rank() {
-    }
-
-    function suit() {
-    }
-
+rank = function() {
+         return Math.floor(this.id/4)+1;
+//  // --> 1..13
+//     // code here...
+ };
+
+suit = function() {
+         return ((this.id%4)+1); // --> 1..4
+//     // code here...
+ };
+
+color = function() {
+       var theSuit=this.suit(); //may be NaN
+         return theSuit && ((theSuit<3)? "red": "black");
+};
+
+cardName = function() {
+         var theRank = this.rank();
+         var theSuit = this.suit();
+         return theRank && theSuit &&
+         (makeCard.rankNames[theRank]+' of '+makeCard.suitNames[theSuit]);
```

Comment:
Line 73 in its current state is causing an error in the cardName function. If you run this function at the moment it will give an error saying that (a number) cannot be determined of property undefined. The undefined property in this case is makeCard, and the reason for this is that your cardName function is actually living in the same level as your rankNames (and suitNames) array. This is what you want, but you don't need to reference the makeCard function for any sort of traversal. just rankNames and suitNames on their own are good enough.